lib/bitmap.rb

```diff
@@ -0,0 +1,87 @@
+class Bitmap
+  attr_reader :pixels, :width, :height
+
+  def initialize(width = 1, height = 1)
+    @width = width.to_i
+    @height = height.to_i
+    @pixels = Array.new(@height)
+    clear
+  end
+
+  def clear
+    @pixels.map! { |_row| WHITE * @width }
+  end
+
+  def paint_pixel(x, y, color)
+    @pixels[y - 1][x - 1] = color.upcase if pixel_exists(x, y)
+  end
+
+  def vertical_segment(x, y1, y2, color)
+    segment_values(y1, y2).each do |y|
+      paint_pixel(x, y, color)
+    end
+  end
+
+  def horizontal_segment(x1, x2, y, color)
+    segment_values(x1, x2).each do |x|
+      paint_pixel(x, y, color)
+    end
+  end
+
+  def fill(x, y, color)
+    matching_color = pixel_color(x, y)
+    pixels_to_fill = [[x, y]]
+
+    until pixels_to_fill.empty?
+      pixel = pixels_to_fill.shift
+
+      paint_pixel(pixel[0], pixel[1], color)
+
+      pixels_to_fill = add_matching_neighboring_pixels(pixels_to_fill, pixel, matching_color)
+    end
+  end
+
+  def render
+    @pixels.join("\n")
+  end
+
+  private
+
+  WHITE = 'O'.freeze
+
+  def pixel_exists(x, y)
```

Comment:
It's considered good practise in Ruby that predicate methods, eg anything that returns a bool, should have a question mark at the end.
```suggestion
  def pixel_exists?(x, y)
```